lib/utils.js

```diff
@@ -39,6 +39,21 @@ exports.pick = function (obj, arr) {
   return res;
 };
 
+// just like emitter.once(event, listener), but doesn't
+// add additional listeners if any listeners with the
+// key are already attached
+exports.oncePerKey = function oncePerKey(emitter, event, key, listener) {
+  var listeners = emitter.listeners(event).filter(function (listener) {
+    return listener._key === key;
+  });
+  if (listeners.length) {
+    return;
+  }
+  emitter.once(event, listener);
+  listeners = emitter.listeners(event);
+  listeners[listeners.length - 1]._key = key;
+};
+
```

Comment:
Actually wait, there is no way to sub out the old listener for a new listener, unless we hold on to a reference to all listeners. Ugh.